rfc/2024-11-13-provider-oci-registry.md

```diff
@@ -0,0 +1,240 @@
+
+There are two possible approaches for distributing providers.
+
+<details>
+<summary>
+
+#### 📂 Standard container tooling approach
+
+</summary>
+
+As a provider author, you may want to package up your provider binary into an OCI image. This involves the following steps:
+
+1. Take all build artifacts and add them to a container image.
+2. Upload the container image to an OCI registry.
+
+For example, assuming you have your binary built and wanted to package it up into a container, you could use the following `Dockerfile`:
+
+```Dockerfile
+FROM scratch
+
+ADD terraform-provider-aws
+ADD README.md
+ADD LICENSE
+ADD CHANGELOG.md
+```
+
+You could now run `docker build -t ghcr.io/opentofu/terraform-provider-aws:1.2.3-amd64 .` followed by `docker push ghcr.io/opentofu/terraform-provider-aws:1.2.3`.
+
+For more advanced use cases, you could enable container image building in your [goreleaser config](https://goreleaser.com/customization/docker/) to enable building multi-arch images.
+</details>
+
+<details>
+<summary>
+
+#### 📂 ORAS approach
+
+</summary>
+
+As a provider author, you may want to package up your provider binary into an OCI image. You can do this by using the [ORAS command line tool](https://oras.land/docs/installation) using the following commands:
+
+```
+oras login -u username -p password ghcr.io
+oras push ghcr.io/opentofu/terraform-provider-aws:1.2.3 ./dist:application/octet-stream+tar
+```
+
+🛑🛑🛑 TODO: how do you create multi-arch with ORAS? 🛑🛑🛑
+
+</details>
+
+### Mirror configuration
+
+If you would like to use an OCI registry as your OpenTofu mirror, you will be able to do so by specifying a `provider_installation` block with an `oci_mirror` section as follows:
+
+```hcl
+provider_installation {
+  oci_mirror {
+    url = "ghcr.io/{{ .Namespace }}/terraform-provider-{{ .Name }}"
+  }
+}
+```
+
+Note that the `url` field is a Go template to resolve the container image address containing the provider binary. You can use `{{ .Namespace }}` to add the provider namespace (e.g. `hashicorp`) and `{{ .Name }}` to add the provider name (e.g. `aws`) to the container image address.
+
+> [!NOTE]
+> OpenTofu will **not** start a container from this image, but instead use the container image as a way to extract the provider binary. Also note that providers are not automatically available in any particular container registry. This feature is intended for advanced use cases.
+
+### The first-class configuration
+
+If you would like to use a provider hosted in an OCI registry, you may do so by invoking a well-known public registry in your `required_providers` block.
+
+```hcl
+terraform {
+  required_providers {
+    aws = {
+      source  = "ghcr.io/opentofu/terraform-provider-aws"
+      version = "~> 1.0"
+    }
+  }
+}
+```
+
+OpenTofu automatically identifies the following registries as OCI registries:
+
+- `docker.io`
+- `ghcr.io`
+- `containers.pkg.github.com`
+- `quay.io`
+- `gcr.io`
+- `public.ecr.aws`
```

Comment:
If we intend to have any special treatment of any hostname that doesn't belong to us, we should seek explicit consent from the owners of that hostname first. We use DNS as the root namespace in provider source addresses specifically so that domain ownership implies control over a particular set of names, and so we shouldn't presume that the owners of these domains will not want to run their own OpenTofu-styled services on them one day.

This sort of conflict already arose for module source addresses where we claimed a hostname owned by an entity that later wanted to run their own module registry, leading to an awkward workaround.

As a compromise, we could reserve some hostnames in a domain we control to act as aliases for each of these registries. I will use `opentofu.org` here as an example, but given how long that is we might consider obtaining a shorter name to use:

* `docker.io.opentofu.org`
* `ghcr.io.opentofu.org`
* `containers.pkg.github.com.opentofu.org`
* `quay.io.opentofu.org`
* `gcr.io.opentofu.org`
* `public.ecr.aws.opentofu.org`

We could still retain the special handling of those hostnames in OpenTofu CLI instead of performing service discovery if we wanted, although that's no longer strictly necessary if we control the hostname.

Registering a new domain to hold all of these aliases would potentially allow us to establish a systematic pattern where _anything_ under that reserved domain name gets projected systematically onto a predefined mapping locally by OpenTofu CLI. Pretending for a moment that the new domain would be `example.com` (which of course it would not really be), perhaps the rule would be:

If the source address's hostname is `anything.example.com`, OpenTofu will construct an OCI repository address using the hostname `example` and the name pattern `${namespace}/terraform-provider-${name}`. For example, `anything.example.com/foo/bar` would be handled as the OCI repository address `anything/foo/terraform-provider-bar` by default.

We could still offer an explicit service-discovery-based approach for those who want to use a different translation for their own OCI registry, but this special hostname would offer a convenience for anyone who is already running a container registry that can handle the addresses generated by that default rule so that they don't need to set up anything special (aside from following the naming scheme) to use their OCI registry with OpenTofu.

Reply:
IMO we can drop this part (specifying OCI source per provider) of the feature altogether and proceed with a generic OCI mirror configuration